cmd/snap-bootstrap/initramfs_systemd_mount.go

```diff
@@ -45,6 +46,132 @@ Wants=%[1]s
 	doSystemdMount = doSystemdMountImpl
 )
 
+// forbiddenChars is a list of characters that are not allowed in any mount paths used in systemd-mount.
+const forbiddenChars = `\,:" `
+
+type fsOpts interface {
+	AppendOptions([]string) ([]string, error)
+}
+
+// overlayFsOptions groups the options to systemd-mount related to overlayfs.
+type overlayFsOptions struct {
+	// Directories to be used as lower layers of an overlay mount.
+	// It does not need to be on a writable filesystem.
+	LowerDirs []string
+	// Optional. A directory to be used as the upper layer of an overlay mount.
+	// This is normally on a writable filesystem.
+	UpperDir string
+	// Optional. A directory to be used as the workdir of an overlay mount.
+	// This needs to be an empty directory on the same filesystem as upperdir.
+	WorkDir string
+}
+
+// validate is used to perform consistency checks on the options related to overlayfs mounts.
+func (o *overlayFsOptions) validate() error {
+	if len(o.LowerDirs) <= 0 {
+		return errors.New("missing arguments for overlayfs mount: at least one lowerdir is required")
+	}
+
+	if len(o.UpperDir) > 0 && len(o.WorkDir) <= 0 {
+		return errors.New("an upperdir for an overlayfs mount was specified but workdir is missing")
+	}
+
+	if len(o.WorkDir) > 0 && len(o.UpperDir) <= 0 {
+		return errors.New("a workdir for an overlayfs mount was specified but upperdir is missing")
+	}
+
+	if strings.ContainsAny(o.UpperDir, forbiddenChars) {
+		return fmt.Errorf("upperdir overlayfs mount option contains forbidden characters. %q contains one of %q", o.UpperDir, forbiddenChars)
+	}
+
+	if strings.ContainsAny(o.WorkDir, forbiddenChars) {
+		return fmt.Errorf("workdir overlayfs mount option contains forbidden characters. %q contains one of %q", o.WorkDir, forbiddenChars)
+	}
+
+	for _, d := range o.LowerDirs {
+		if strings.ContainsAny(d, forbiddenChars) {
+			return fmt.Errorf("lowerdir overlayfs mount option contains forbidden characters. %q contains one of %q", d, forbiddenChars)
+		}
+	}
+
+	return nil
+}
+
+// AppendOptions constructs the overlayfs related arguments to systemd-mount after validation.
+func (o *overlayFsOptions) AppendOptions(options []string) ([]string, error) {
+	err := o.validate()
+	if err != nil {
+		return nil, err
+	}
+
+	var lowerDirs strings.Builder
+	for i, d := range o.LowerDirs {
+		// This is used for splitting multiple lowerdirs as done in
+		// https://elixir.bootlin.com/linux/v6.10.9/C/ident/ovl_parse_param_split_lowerdirs.
+		if i != 0 {
+			lowerDirs.WriteRune(':')
+		}
+
+		lowerDirs.WriteString(d)
+	}
```

Comment:
Can this be a `lowerDirs := strings.Join(o.LowerDirs, ":")`

Reply:
doh! ofc it can. This piece of code changed so many times since I introduced it that I completely missed this. Thanks!